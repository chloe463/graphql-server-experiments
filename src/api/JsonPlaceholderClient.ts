import axios, { AxiosInstance } from "axios";
import faker from "faker";
import { NexusGenObjects } from "../graphql/generated/typings";

type Post = NexusGenObjects["Post"];
type Comment = NexusGenObjects["Comment"];

const BASE_URL = "https://jsonplaceholder.typicode.com/";

type FetchPostsParams = {
  start?: string;
  limit?: number;
  page?: number;
};

type FetchPostsResponse = {
  posts: Post[];
  totalCount: number;
};

type FetchCommentsResponse = {
  comments: Comment[];
};

export class JsonPlaceholderClient {
  private client: AxiosInstance;
  constructor() {
    this.client = axios.create({
      baseURL: BASE_URL,
    });
  };

  fetchPosts = async (args: FetchPostsParams): Promise<FetchPostsResponse> => {
    const params = {};
    Object.keys(args).forEach((key) => {
      params[`_${key}`] = args[key];
    });
    const res = await this.client.get<Post[]>("/posts", {
      params: {
        ...params,
      }
    });
    return { posts: res.data, totalCount: res.headers["x-total-count"] };
  };

  makeDummyPosts = (args: FetchPostsParams): FetchPostsResponse => {
    const posts = Array.from({ length: args.limit || 10 }, (_, i) => i + 1).map((i) => {
      const id = parseInt(args.start || "0", 10) + i;
      return {
        id,
        title: faker.lorem.word(10),
        body: faker.lorem.words(100),
        userId: Math.ceil(id / 10),
      };
    });
    return { posts, totalCount: 100 };
  }

  fetchComments = async (postId: number): Promise<FetchCommentsResponse> => {
    const res = await this.client.get<Comment[]>(`/posts/${postId}/comments`);
    return { comments: res.data };
  }

  makeDummyComments = (postId: number): FetchCommentsResponse => {
    const comments: Comment[] = Array.from({ length: 5 }, (_, i) => i + 1).map((i) => {
      const id = (postId - 1) + i;
      return {
        id,
        body: faker.lorem.words(100),
        email: `${faker.name.firstName()}@example.com`,
        name: faker.name.findName(),
        postId,
      }
    });

    return { comments };
  }
}
