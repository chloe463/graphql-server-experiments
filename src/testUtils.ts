import { ApolloServer } from "@apollo/server";
// import typeDefs from "./graphql/generated/schema.graphql";
import { schema } from "./graphql/schema";

export const constructTestServer = () => {
  const server = new ApolloServer({
    schema,
  });

  return server;
};
