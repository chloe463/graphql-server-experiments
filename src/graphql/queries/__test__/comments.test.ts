import { gql } from "apollo-server";
import { JsonPlaceholderClient } from "../../../api/JsonPlaceholderClient";
import { constructTestServer } from "../../../testUtils";

const jsonPlaceholderClientMock = (JsonPlaceholderClient as any) as jest.Mock<JsonPlaceholderClient>;

const GET_COMMENTS_QUERY = gql`
  query GetCommentsQuery($postId: Int!) {
    comments(postId: $postId) {
      id
      postId
      name
      email
      body
    }
  }
`;

describe("[Query] comments", () => {
  it("returns comments", async () => {
    (jsonPlaceholderClientMock as any).makeDummyComments = jest.fn().mockReturnValue({
      comments: [
        {
          id: 1,
          postId: 1,
          name: "id labore ex et quam laborum",
          email: "Eliseo@gardner.biz",
          body: "laudantium enim quasi est quidem magnam voluptate ipsam eos\ntempora quo necessitatibus\ndolor quam autem quasi\nreiciendis et nam sapiente accusantium"
        },
        {
          id: 2,
          postId: 1,
          name: "quo vero reiciendis velit similique earum",
          email: "Jayne_Kuhic@sydney.com",
          body: "est natus enim nihil est dolore omnis voluptatem numquam\net omnis occaecati quod ullam at\nvoluptatem error expedita pariatur\nnihil sint nostrum voluptatem reiciendis et"
        },
        {
          id: 3,
          postId: 1,
          name: "odio adipisci rerum aut animi",
          email: "Nikita@garfield.biz",
          body: "quia molestiae reprehenderit quasi aspernatur\naut expedita occaecati aliquam eveniet laudantium\nomnis quibusdam delectus saepe quia accusamus maiores nam est\ncum et ducimus et vero voluptates excepturi deleniti ratione"
        },
        {
          id: 4,
          postId: 1,
          name: "alias odio sit",
          email: "Lew@alysha.tv",
          body: "non et atque\noccaecati deserunt quas accusantium unde odit nobis qui voluptatem\nquia voluptas consequuntur itaque dolor\net qui rerum deleniti ut occaecati"
        },
        {
          id: 5,
          postId: 1,
          name: "vero eaque aliquid doloribus et culpa",
          email: "Hayden@althea.biz",
          body: "harum non quasi et ratione\ntempore iure ex voluptates in ratione\nharum architecto fugit inventore cupiditate\nvoluptates magni quo et"
        }
      ],
    });
    const server = constructTestServer({
      jsonPlaceholderClient: jsonPlaceholderClientMock,
    });

    const res = await server.executeOperation({ query: GET_COMMENTS_QUERY, variables: { postId: 1 } });
    expect(res).toMatchSnapshot();
  });
});
