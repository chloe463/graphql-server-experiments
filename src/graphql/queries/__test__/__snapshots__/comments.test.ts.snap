// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Query] comments returns comments 1`] = `
Object {
  "body": Object {
    "kind": "single",
    "singleResult": Object {
      "data": Object {
        "comments": Array [
          Object {
            "body": "laudantium enim quasi est quidem magnam voluptate ipsam eos
tempora quo necessitatibus
dolor quam autem quasi
reiciendis et nam sapiente accusantium",
            "email": "Eliseo@gardner.biz",
            "id": 1,
            "name": "id labore ex et quam laborum",
            "postId": 1,
          },
          Object {
            "body": "est natus enim nihil est dolore omnis voluptatem numquam
et omnis occaecati quod ullam at
voluptatem error expedita pariatur
nihil sint nostrum voluptatem reiciendis et",
            "email": "Jayne_Kuhic@sydney.com",
            "id": 2,
            "name": "quo vero reiciendis velit similique earum",
            "postId": 1,
          },
          Object {
            "body": "quia molestiae reprehenderit quasi aspernatur
aut expedita occaecati aliquam eveniet laudantium
omnis quibusdam delectus saepe quia accusamus maiores nam est
cum et ducimus et vero voluptates excepturi deleniti ratione",
            "email": "Nikita@garfield.biz",
            "id": 3,
            "name": "odio adipisci rerum aut animi",
            "postId": 1,
          },
          Object {
            "body": "non et atque
occaecati deserunt quas accusantium unde odit nobis qui voluptatem
quia voluptas consequuntur itaque dolor
et qui rerum deleniti ut occaecati",
            "email": "Lew@alysha.tv",
            "id": 4,
            "name": "alias odio sit",
            "postId": 1,
          },
          Object {
            "body": "harum non quasi et ratione
tempore iure ex voluptates in ratione
harum architecto fugit inventore cupiditate
voluptates magni quo et",
            "email": "Hayden@althea.biz",
            "id": 5,
            "name": "vero eaque aliquid doloribus et culpa",
            "postId": 1,
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": Object {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
